movement component for mechanics

zjdtmkhzt opened the pull request:
<!-- The text between the arrows are comments - they won't be visible on your PR. -->
<!-- To label this PR, add the label(s) without the prefixes surrounded by brackets anywhere, for example: [LABEL] -->
<!-- PRs should at least have one area (A-) label and at least one category (C-) label -->

## About the PR <!-- Describe the Pull Request here. What does it change? What other things could this impact? -->
This adds a new mechanics component that can only be used inside device frames.
It allows the device frame to move around under its own power.

Directions can be given as words ("north") or as byond numeric directions ("1").
The walking speed is configurable, the maximum should be normal walking speed.
A frame can be made to either take one step, or to continue walking until the command is cancelled by giving another.

The component also makes the frame grow terrifying looking badly sprited legs.
This choice was made due to my own lack of artistic skill, if anyone wants to make some kind of leg/wheel/thruster sprites, please let me know!

Side things:
I also changed the cabinets to use a medium light instead of a complex one, due to them possibly moving more in the future.
I made two new leg component subtypes and made the original a bit more generic.
I added a proc for components that is called when their cabinet is wrenched/unwrenched from the floor.
I added a field for components to specify if they can only be used inside cabinets.

Here is me addressing a frame via packets from my PDA:

https://user-images.githubusercontent.com/35579460/198901636-3eede7b2-6791-472b-8002-e27ea387eed6.mp4

## Why's this needed? <!-- Describe why you think this should be added to the game. -->
- I think this could be a fun addition to allow mechanics to make interesting contraptions! 
- There's not really anything bad you can do with cabinets, since things like gravitons, gun components, tesla coils, etc... are all blacklisted from them already.
- Hopefully in the future me or someone else can add other components which can act as sensors, like for instance a GPS component to get your (or other GPS units) position, or some kind of component to let you sense nearby walls/mobs/objects?


## Changelog <!-- If necessary, put your changelog entry below. Otherwise, /please/ delete this entire section. -->
<!-- Put how you want to be credited in the changelog in place of CodeDude. -->
<!-- Use (*) for major changes and (+) for minor changes. See the contributor guide for details. For example: -->

```changelog
(u) zjdtmkhzt
(+) Added a new mechanics component that lets device frames move around. You can read about it in the ingame mechanics book.
```

Conchuckter commented:
Awesome idea!

It'd be great to have some sensing so we can get information about the cabinets surroundings though since in its current state it could only really be programmed or directly controlled.

pali6 commented:
> It'd be great to have some sensing so we can get information about the cabinets surroundings though since in its current state it could only really be programmed or directly controlled.

This is planned. And by planned I mean that when talking about this PR we all agreed that it'd be a good idea but no one has confirmed they will acutally code it. But I'll do my best to bully zjdt into making those PRs too.

pali6 commented on code/modules/mechanics/MechanicMadness.dm:
```diff
@@ -3291,7 +3293,120 @@
+			src.loc.AddComponent(/datum/component/legs/four)
+		else
+			src.loc.AddComponent(/datum/component/legs/six)
+		RegisterSignal(src.loc, COMSIG_MOVABLE_SET_LOC, .proc/container_check)
```
Should this really happen even if our loc is a turf?

pali6 replied:
Missed the cabinet_only var.

pali6 commented on code/modules/mechanics/MechanicMadness.dm:
```diff
@@ -3291,7 +3293,120 @@
+		if (istype(src.loc, /obj/item/storage/mechanics/housing_handheld))
+			src.loc.AddComponent(/datum/component/legs/four)
+		else
+			src.loc.AddComponent(/datum/component/legs/six)
```
Same here.

pali6 commented on code/modules/mechanics/MechanicMadness.dm:
```diff
@@ -3291,7 +3293,120 @@
+		if (!istype(S))
+			return
+		if (!isturf(S.loc))
+			stop_moving()
```
I'm not sure what is desired behaviour but should moving return when we are on a turf again?

If so you will likely need to store last walk arguments.

zjdtmkhzt replied:
No, I think it's fine if it stops!

zjdtmkhzt replied:
Intended behavior for me is to have it stop while in a container, so I think this is fine!

pali6 commented on code/modules/mechanics/MechanicMadness.dm:
```diff
@@ -3291,7 +3293,120 @@
+		if (container.anchored)
+			src.stop_moving()
+
+	proc/do_walk(var/datum/mechanicsMessage/input)
```
Both walk and step should check if S.loc is a turf before they activate.

pali6 commented on code/modules/mechanics/MechanicMadness.dm:
```diff
@@ -3291,7 +3293,120 @@
+		if (direction == 0)
+			UnregisterSignal(S, COMSIG_MOVABLE_MOVED)
+		else
+			RegisterSignal(S, COMSIG_MOVABLE_MOVED, .proc/set_glide_size, TRUE)
```
I kinda hate this. But I guess there's no avoiding it since pulling the thing while it walks would break glide settings, right?

Yet another reason why `walk` sucks btw.

zjdtmkhzt replied:
yeah, I think it has to be this way pensive emoji

pali6 commented on code/modules/mechanics/MechanicMadness.dm:
```diff
@@ -3291,7 +3293,120 @@
+		set_glide_size()
+		step(S, direction, (32 / move_lag) * world.tick_lag)
+		set_glide_size()
+		if (direction == 0)
+			UnregisterSignal(S, COMSIG_MOVABLE_MOVED)
+		else
+			RegisterSignal(S, COMSIG_MOVABLE_MOVED, .proc/set_glide_size, TRUE)
```
For step this shouldn't be necessary, right? It should probably just unregister (because it cancels current walk).

zjdtmkhzt replied:
good point!

zjdtmkhzt commented:
Alright, I added more checks for if the frames are inside things, or in space, so they should now stop (or not start) if that is the case.

I also fixed conveyor code so that it should no longer cancel the walking of things that are supposed to be unaffected by conveyors (and I also applied this property to moving frames).
